server-test/Hrafnar/ParserSpec.hs

```diff
@@ -17,6 +17,154 @@ spec :: Spec
+
+          parseExpr "42"
+          `shouldParse`
+          Lit' (Int' 42)
+
+      context "character" $ do
+
+        context "without escaping" $ do
+
+          it "parse a character" $
+
+            parseExpr "'c'"
+            `shouldParse`
+            Lit' (Char' 'c')
+
+          it "parse a numeric character" $
+
+            parseExpr "'5'"
+            `shouldParse`
+            Lit' (Char' '5')
+
+          it "parse a space" $
+
+            parseExpr "' '"
+            `shouldParse`
+            Lit' (Char' ' ')
+
+          it "fail on a backslash" $
+
+            parseExpr "'\\'"
+            `shouldFailWith`
+            err 3 (ueof <> etok '\'') -- This error means no ending quote mark.
+
+          it "parse a single quote" $
+
+            parseExpr "'''"
+            `shouldParse`
+            Lit' (Char' '\'')
+
+          it "parse a double quote" $
+
+            parseExpr "'\"'"
+            `shouldParse`
+            Lit' (Char' '\"')
+
+          it "fail on a zero length character" $
+
+            parseExpr "''"
+            `shouldFailWith`
+            err 2 (ueof <> etok '\'') -- This error means no ending quote mark.
+            -- NOTE: This error depends on whether HML tries to parse a single
+            -- quote without escaping or not.
+            -- If it did not, the error might be:
+            --     err 1 (utok '\'')
+
+          it "fail on a multiple length character" $
+
+            parseExpr "'ab'"
+            `shouldFailWith`
+            err 2 (utok 'b' <> etok '\'')
+
+        context "with escaping" $ do
+
+          it "parse a special character" $
+
+            parseExpr "'\\n'"
+            `shouldParse`
+            Lit' (Char' '\n')
+
+          it "parse a decimal unicode" $
+
+            parseExpr "'\\74'"
+            `shouldParse`
+            Lit' (Char' 'J')
+
+          it "parse a hexadecimal unicode" $
+
+            parseExpr "'\\x4B'"
+            `shouldParse`
+            Lit' (Char' 'K')
+
+          it "parse a octal unicode" $
+
+            parseExpr "'\\o114'"
+            `shouldParse`
+            Lit' (Char' 'L')
+
+          it "parse a ascii control code abbreviation" $
+
+            parseExpr "'\\LF'"
+            `shouldParse`
+            Lit' (Char' '\n')
+
+          it "parse a caret notation" $
+
+            parseExpr "'\\^J'"
+            `shouldParse`
+            Lit' (Char' '\n')
+
+      context "string" $ do
+
+        it "parse a string which length is 0" $
+
+          parseExpr "\"\""
+          `shouldParse`
+          Lit' (String' "")
+
+        it "parse a string which length is 1" $
+
+          parseExpr "\"s\""
+          `shouldParse`
+          Lit' (String' "s")
+
+        it "parse a string which length is 2 or over" $
+
+          parseExpr "\"st\""
+          `shouldParse`
+          Lit' (String' "st")
+
+        it "parse a string including escaped characters" $
+
+          parseExpr "\" \\n \\74 \\' \\\" \\\\ \""
```

Comment:
This fails to the complicated representation since there are two kind backslashes: to escape for Haskell and to does for HML.
In a new issue, I will improve it with some raw string representations.

Reply:
I've created the issue #56 .